### PR TITLE
Size block

### PR DESCRIPTION
**Pull request by nkiruka**
This PR covers the addition of a new size block with a size input and sticker dropdown. 
Previously, users did not have the option to input a custom size as shown in "initial option" below.  The addition of the size block to the toolbox, gives users the option to input a custom size to enhance user experience. 

Initial option
prework-stickersize

This is the implementation of the new size block that allows a user to input any numerical value for size.  The sticker drawn is resized per the size input.

post-stickersizev2

Artist toolbox that includes two sticker options
sticker-rev1

### CODE REVIEW

File: apps/src/turtle/turtle.js [hunk 1]
```diff
@@ -1585,7 +1585,7 @@ Artist.prototype.step = function (command, values, options) {
       this.visualization.ctxScratch.save();
       this.visualization.ctxScratch.translate(this.visualization.x, this.visualization.y);
       this.visualization.ctxScratch.rotate(this.visualization.degreesToRadians_(this.visualization.heading));
-      this.visualization.ctxScratch.drawImage(img, 0, 0, width, height, -width / 2, -height, width, height);
+      this.visualization.ctxScratch.drawImage(img, -width / 2, -height, width, height);
```
Comment (joshlory) [inline apps/src/turtle/turtle.js]: I repro the `IndexSizeError: DOM Exception 1` with the following code:

```
phantomjs> img = new Image(100, 100);
phantomjs> ctx = document.createElement('canvas').getContext('2d');
phantomjs> ctx.drawImage(img, 0, 0, 100, 100);
IndexSizeError: DOM Exception 1

  phantomjs://repl-input:1 in global code
```

Can we continue to pass all 9 arguments here to fix the failing test?

File: apps/src/turtle/blocks.js [hunk 2]
```diff
@@ -1129,6 +1129,39 @@ exports.install = function (blockly, blockInstallOptions) {
         '", null, \'block_id_' + this.id + '\');\n';
   };
 
+  blockly.Blocks.turtle_sticker_with_size = blockly.Blocks.turtle_stamp = {
```
Comment (joshlory) [inline apps/src/turtle/blocks.js]: Why also export as `turtle_stamp`?

File: apps/src/turtle/blocks.js [hunk 3]
```diff
@@ -1129,6 +1129,39 @@ exports.install = function (blockly, blockInstallOptions) {
...
+      // Generates a list of pairs of the form [[url, name]]
+      var values = [];
+      for (var name in skin.stickers) {
+        var url = skin.stickers[name];
+        values.push([url, name]);
+      }
+
+      dropdown = new blockly.FieldImageDropdown(values, 40, 40);
+
+      input.appendTitle(dropdown, 'VALUE');
```
Comment (joshlory) [inline apps/src/turtle/blocks.js:1149]: I wonder if we can share the code from line 1134-1149 and 1152-1155 with `blockly.Blocks.sticker` above?

File: apps/src/turtle/blocks.js [hunk 4]
```diff
@@ -1129,6 +1129,39 @@ exports.install = function (blockly, blockInstallOptions) {
...
+      this.setPreviousStatement(true);
+      this.setNextStatement(true);
+      this.setTooltip(msg.drawSticker());
+      this.setTooltip(msg.setStickerSize());
+    }
+  };
+
+  generator.turtle_sticker_with_size = generator.turtle_stamp = function () {
+    var size = generator.valueToCode(this, 'SIZE', Blockly.JavaScript.ORDER_NONE);
+    return 'Turtle.drawSticker("' + this.getTitleValue('VALUE') + '", ' + size + ', \'block_id_' + this.id + '\');\n';
```
Comment (joshlory) [inline apps/src/turtle/blocks.js]: Nit: this is a perfect use case for [string interpolation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) 😄.

Comment (nkiruka) [reply apps/src/turtle/blocks.js]: I concur!

File: apps/src/turtle/blocks.js (hunk 2 above)
Comment (nkiruka) [reply apps/src/turtle/blocks.js]: Redundant!  Thanks for spotting.

File: apps/src/turtle/blocks.js (hunk 3 above)
Comment (nkiruka) [reply apps/src/turtle/blocks.js:1149]: Deleted "turtle_stamp" from lines 1132 & 1160.

**Comment by nkiruka** [comment]
Updated feedback.  Please take a look.

**Comment by Hamms** [comment]
"Set sticker size" implies to me that it's changing the size of the sticker drawn by the "draw sticker" block. I'd suggest we instead label this block something like `draw sticker <sticker dropdown> at size <size input>`

File: apps/src/turtle/blocks.js [hunk 5]
```diff
@@ -1129,6 +1129,39 @@ exports.install = function (blockly, blockInstallOptions) {
...
+      dropdown = new blockly.FieldImageDropdown(values, 40, 40);
+
+      input.appendTitle(dropdown, 'VALUE');
+      this.appendValueInput('SIZE');
+
+      this.setInputsInline(true);
+      this.setPreviousStatement(true);
+      this.setNextStatement(true);
+      this.setTooltip(msg.drawSticker());
+      this.setTooltip(msg.setStickerSize());
```
Comment (Hamms) [inline apps/src/turtle/blocks.js]: Does setting two tooltips make sense, or will the second override the first?

File: apps/test/unit/turtle/turtleTest.js [hunk 6]
```diff
@@ -105,6 +105,72 @@ describe('Artist', () => {
...
+      let blockId = "block_id_4";
+      let options = {smoothAnimate: false};
+
+      artist.visualization = new Artist.Visualization();
+      const setStickerSize = sinon.spy(artist.visualization.ctxScratch, 'drawImage');
+      artist.stickers = {Alien:img};
+      artist.step('sticker', ['Alien', size, blockId], options);
+
+      expect(setStickerSize).to.be.have.been.calledWith(img, 0, 0, 100, 40, -15, -12, 30, 12);
+
```
Comment (Hamms) [inline apps/test/unit/turtle/turtleTest.js:173]: great tests!

Comment (nkiruka) [reply apps/test/unit/turtle/turtleTest.js:173]: Thanks!

**Comment by nkiruka** [comment]
Updated PR, please review.

**Pull request merged**